src/walker.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Pipeline;
+  stepFn?: (step: Step) => Step;
+  commandFn?: (command: Command) => Command;
+}
+
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  const pipeline = new Pipeline(p.name);
+  pipeline.steps = [];
+  for (const step of p.steps) {
+    let newStep: PotentialStep | null = cloneDeep(step);
+    newStep = await evaluateStep(newStep, conditionalCache);
+    if (newStep) {
+      pipeline.steps.push(newStep);
+    }
+  }
+  return pipeline;
+}
+
+async function evaluateStep(
+  step: PotentialStep,
+  conditionalCache: Map<any, any>,
+): Promise<Step | null> {
+  if (step instanceof Conditional) {
+    if (await step.accept()) {
+      step = await getAndCacheDependency(conditionalCache, step);
+    } else {
+      return null;
+    }
+  }
+  const deps = step.dependencies;
+  const newDeps: Set<Step> = new Set();
+  for (const dep of deps) {
+    const newDep = await evaluateStep(dep, conditionalCache);
+    if (newDep) {
+      newDeps.add(newDep);
+    }
+  }
+  step.dependencies = newDeps;
+  return step;
+}
+
+export async function walk(p: Pipeline, mutator: Mutators): Promise<Pipeline> {
```

Comment:
Unless the mutators are async there's no need for this function to be async. Either change this function or update the mutators' signature.

Reply:
changed the mutators to return promises